learn/components/styling.md

```diff
@@ -3,4 +3,309 @@ title: Styling
+
+### Writing default styles for the shadow host with `:host` and `:host()`
+
+There are two selectors which can be used to style the shadow host from within the shadow tree. They are the `:host`
+pseudo-class and the `:host()` pseudo-class function. The first will always select the shadow host. Here’s `:host` in
+action:
+
+```html
+<fancy-p>
+  <template shadowroot="open">
+    <style>
+      :host {
+        display: inline-block;
+      }
+    </style>
+  </template>
+</fancy-p>
+```
+
+The second will select the shadow host if it matches a selector. For example, given the following default styles for the
+`<fancy-p>` component, only the second element in the subsequent code example will have the styles applied.
+
+```css
+:host([extra]) {
+  font-style: italic;
+  font-weight: bold;
+}
+```
+
+```html
+<fancy-p>I not am extra</fancy-p>
+<fancy-p extra>I am extra</fancy-p>
+```
+
+#### Chaining selectors after `:host`
+
+While the `:host` selector refers to the shadow host element which is outside of the shadow tree, if you chain selectors
+it will select elements within the shadow tree.
+
+```html
+<fancy-p>
+  <template shadowroot="open">
+    <style>
+      :host > p {
+        color: deeppink;
+      }
+    </style>
+    <p>I am deeppink.</p>
+  </template>
+  <p>I am not deeppink.</p>
+</fancy-p>
+```
+
+You can `::slotted()` to style light DOM descendants of a shadow host.
+
+### Writing default styles for slotted elements with `::slotted()`
+
+The `::slotted()` pseudo-element selector allows you to write default styles for different kinds of elements which are
+slotted.
+
+```html
+<fancy-elements>
+  <template shadowroot="open">
+    <style>
+      ::slotted(button) {
+        color: deeppink;
+      }
+
+      ::slotted(p) {
+        color: teal;
+      }
+    </style>
+    <slot></slot>
+  </template>
+  <button>I am a slotted button</button>
+  <p>I am a slotted paragraph.</p>
+</fancy-elements>
+```
+
+If you want to target elements in specific slots you can pass an attribute selector which matches the slot:
+
+```html
+<fancy-article>
+  <template shadowroot="open">
+    <style>
+      ::slotted([slot="title"]) {
+        font-size: 2rem;
+      }
+      ::slotted([slot="subtitle"]) {
+        font-size: 1.25rem;
+      }
+      /* The following will target elements going into the unnamed slot */
+      ::slotted(:not([slot])) {
+        color: deeppink;
+      }
+    </style>
+    <article>
+      <hgroup>
+        <slot name="title"></slot>
+        <slot name="subtitle"></slot>
+      </hgroup>
+      <slot></slot>
+    </article>
+  </template>
+  <h1 slot="title">I am the title</h1>
+  <p slot="subtitle">I am the subtitle</p>
+  <p>I am content.</p>
+</fancy-article>
+```
+
+You cannot chain selectors with `::slotted()`, so the following will not work:
+
+```css
+::slotted(h1) span {
+  color: deeppink;
+}
+```
+
+## Parts: styling a shadow tree from the outside
```

Comment:
I definitely think Parts deserve their own section, especially as by this point I think we're close to 10 minutes of reading time.

Reply:
Here’s a thought I just had for how to structure all the styling content maybe in different section:

- Introduction
  - Focuses on encapsulation and then gives a brief overview with links to each of the following sections.

- Ways to include styles in the Shadow DOM
  - A necessary but boring prerequisite for everything.

- How to write default styles for Web Components.
  - Covers `:host` and `::slotted`.
  - Shows how to make Web Component Styling portable.

- How to create a styling API for Web Components.
  - Covers using custom properties, `::part()`, `exportparts` for nested custom elements, and maybe even `!important` to limit styling certain properties.
  - Shows how to make Web Component Styling extensible.

There is more to cover like gotchas around `@keyframes` and `@font-face`. Maybe those could go in the “default styles” section.

Reply:
I really like it. I like the idea of layering knowledge. What's the first thing that someone needs to just get styles working? Armed with that knowledge up front, they can experiment, and learn things independently, rather than having to read the entirety. Perhaps the basics can be expanded upon. Talk more about how it could be preferable to simplify selectors given the encapsulation. Maybe the introduction and "ways to include..." can be combined into a "Here's the basics" section.

Parts, while foundational, are more about architecting a component. So we should batch that with other architectural parts. So yeah I think having a section about styling API & extensibility is great.